pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -121,6 +126,8 @@ async def maybe_ack(ack_id: str) -> None:
                 metrics_client.increment('pubsub.acker.batch.failed')
 
                 continue
+            except asyncio.CancelledError:  # pylint: disable=try-except-raise
+                raise
```

Comment:
Hmm, I think we might be better off implementing graceful shutdown flow in `acker` worker.

Otherwise with this changes `ack_window` will punish consumer for being too performant. Consider ack_queue of `length=10` and `ack_window=30`. No matter how fast producer is, acker will not ack more than 10 messages per 30 seconds.

Alternatively we can also hardcode ack queue length to 2500, but that will require us to check `ack_ids` list length after each message pulled from the queue. But that might not interfere nicely with "inherited" batch logic...

Reply:
>Hmm, I think we might be better off implementing graceful shutdown flow in acker worker.

I agree that right now, the shutdown logic for acker is wrapped up in the `consumer`, but I thought this change was the most straight forward extension of the current logic since we're already joining the acker queue during consumer shutdown. I suppose we could change this and wrap the main worker loop of acker in a separate coroutine and shield it from cancellation as well. Thoughts?

> Otherwise with this changes ack_window will punish consumer for being too performant. Consider ack_queue of length=10 and ack_window=30. No matter how fast producer is, acker will not ack more than 10 messages per 30 seconds.

How is consumer being punished any more with these changes than it was before? I just changed where we were signalling finished tasks, which should only affect when `join.()`ing on the ack queue, which is only during shutdown. Or are you saying this will negatively impact shutdown?

Reply:
> How is consumer being punished any more with these changes than it was before? I just changed where we were signalling finished tasks, which should only affect when .join()ing on the ack queue, which is only during shutdown. Or are you saying this will negatively impact shutdown?

It will also be blocked on `queue.put(...)` if the queue size limit is reached? It seems quite unlikely with proper setup, but with `num_producers=1` and `max_messages_per_producer=6` 🌝  we will be doomed to ack no more than 6 messages per `ack_window`.

>I suppose we could change this and wrap the main worker loop of acker in a separate coroutine and shield it from cancellation as well. Thoughts?

I was thinking about just doing a final `await subscriber_client.acknowledge(subscription, ack_ids=sack_ids)` in `except asyncio.CancelledError` block, but shielding also sounds reasonable. It just a bit harder for me to understand without testing as you can see by my other questions :D

Reply:
> It will also be blocked on queue.put(...) if the queue size limit is reached?

`.put()` seems to be unaffected by `.task_done()`. i.e. something like

```python
q = asyncio.Queue(maxsize=1)
await q.put(1)
await q.get()
await q.put(1)
await q.get()
q.task_done()
q.task_done()
await q.join()
```

succeeds just fine. I think that is where the confusion lies? Or I could just be confused 😄

Reply:
Oh, I see now. Good call, sorry for confusion. Yeah, I believe in that case you initial approach is the most straightforward